### DISCUSSION TITLE
What is right way to create a feature when there are missing values

### DISCUSSION THREAD

**Discussion by gengyabc** [biolab/orange3#6053] [Q&A] [answered]
I want to construct a feature that should be rounded like below, but I "cannot convert float NaN to integer". 

![image](https://user-images.githubusercontent.com/12781437/177553600-d95cd450-a942-423e-9274-a5ec805c9f96.png)

but if I use "Preprocess", I cannot construct my feature

![image](https://user-images.githubusercontent.com/12781437/177554307-67211edb-8da8-4e49-bcb0-afc2239c34c9.png)

Can I use "Test and Score" in this situation? Or what should I do to prevent testing set leaking to training set?

BTW, is below one of right way?
![image](https://user-images.githubusercontent.com/12781437/177555733-7221f7af-5b7e-427b-b655-0ef80a5530b8.png)

**Comment by janezd** [comment]
The error message looks strange ... but can your MonthlyCharges equal 0? Does `round(TotalCharges/MonthlyCharges) if MonthlyCharges > 0 else 0` work better?

Note that you probably don't want to use `0`, as above. Perhaps you should better use Select Rows to remove any rows with zero monthly charges.

**Reply by ajdapretnar** [reply]
That's not the case. I've tried it on heart-disease and the two error messages are different. `ZeroDivisionError` for dividing by 0 and `ValueError` when missing values are in the data. I wasn't able to write a normal-looking statement that would handle both.

**Comment by janezd** [comment]
`,,, if monthlyCharges == monthlyCharges != 0 ... ` is `False` when `monthlyCharges` is either `nan` or `0`.  :)

**Reply by gengyabc** [reply]
There is no 0 in MonthlyCharges.  
I want to impute the missing values before create the new feature but I don't find the proper way to do it if I want to use "Test and Score".

The similar situation is what if I want to use tree based imputer? "Preprocess" widget cannot do it.  The only way is using "Data sampler" with "Prediction"?

In these situations, is there any way to use "Test and Score" ?

**Reply by janezd** [reply]
You can connect Preprocess directly to Test and Score. Preprocessors will be "trained" on training data only, and the used on testing data.

For instance, if you impute mean values, the mean will be computed from training data only, but imputation will also be done on testing data. This is Ok, because no information leaks from testing data to the model.

**Reply by gengyabc** [reply]
But how can I impute before "feature constructor"?

**Comment by janezd** [comment, marked as answer]
Ah, of course, I see.

Can you try using `round(TotalCharges/MonthlyCharges) if MonthlyCharges == MonthlyCharges > 0 else 0` in Feature Constructor? This should handle missing values, I guess.

**Reply by gengyabc** [reply]
yes, this solution kind of solved the problem, though not the ideal way. Thank you very much.
